### ISSUE TITLE
Change Request: Add a config post-processing/validation system

### ISSUE THREAD

**Issue by matwilko** [eslint/eslint#17428]
### ESLint version

8.45.0

### What problem do you want to solve?

There's currently no extension points in the configuration system that allow third-parties to inspect and/or modify configuration in a consistent way that would allow new forms of functionality that would currently require modifications to the core.

It's possible to shim some post-processing onto the config array that will be passed to ESLint by just manually adding that processing at the end of `eslint.config.js`, but that requires the user to explicitly write that support in, and it requires that code to understand and/or reimplement all the logic in config-array/object-schema in order to properly inspect/modify it.

There's a number of feature requests currently open/rejected that I think could reasonably be implemented by community projects rather than ESLint itself if a small but powerful extension point is added. Some examples:
 - Overriding severities/treat warnings as errors/silence warnings, from #14679
 - Clearing out 'off' rules, from #17361
 - Disabling `warn` rules when `--quiet`, from #16450 / #17274
 - Disabling all rules of a specific plugin, from #15216
 - Disable all rules for particular extensions, from #17360

A new extension point could also provide a means for a new kind of extended functionality from the community - "linting" configuration itself. e.g. they could inspect the rules users have configured and potentially warn/error about conflicts between different plugins; rules that don't apply given their ECMAScript level; warn about upcoming deprecations for certain options; companies could ensure that individual teams aren't turning off rules that are required by policy, and probably more scenarios that there's not really a way to support at the moment except through documentation.

### What do you think is the correct solution?

I'd propose adding a new top-level configuration key, `configProcessor`, which would be an array of processors that would merge by concatenating the arrays. `configProcessors` would not appear in the finalized file config, in the same way that `files` and `ignores` aren't included atm.

The config processors would be run during the `finalize` phase for each individual file, operating on the merged configuration for the file.

There would be two extension points for processors, `update` and `validate`.

The `update` extension point is passed the merged (and possibly frozen?) configuration for a file (along with the filePath and possibly the ESLint options?). The function can return one of four things:
 - `undefined`/`null` - I don't need to make any changes this cycle
 - `Symbol('complete')` - I won't need to make any further changes in future cycles
 - `object` - Completely replace the file's config with this new config object
 - `object[]` or `Iterator<object>` - new config array elements to merge into the existing file config (an empty array/iterator is treated as if the method had returned `undefined`)

In both of the cases where the processor updates the config, the config is validated in the normal way to ensure the processors aren't making invalid changes.

The array of processors would be called in order, with each successive processor passed the up-to-date config as potentially modified by previous processors. If _any_ processor changes the config, then the processors are run again in order until all return `undefined` or disable themselves with `Symbol('complete')`. This is to ensure that the processors can always get the config to a stable and predictable state despite other processors not necessarily doing things the same exact way (obviously there'll need to be a limit on the number of cycles this can go on for, as processors might conflict and infinitely change the config back and forth)

After all the processors have made any updates they need to, the _final_ file config is then validated by the processors. The `validate` extension point is called on each processor in order, and can return an array/iterator of objects that detail validation errors in the config (location(s), message, etc.)

So as an example, "treat warnings as errors" could be implemented like so:

```js
const treatWarningsAsErrors = {
    update*(filePath, fileConfig, eslintOptions) {
        // Just assuming that the rules will be in their canonical [severity, ...] form for simplicity here
        const warningRules = Object.keys(fileConfig.rules).filter(ruleName => fileConfig.rules[ruleName][0] === 1);
        yield return { rules: Object.fromEntries(warningRules.map(ruleName => [ruleName, 2])) };
    }
};
```

As another example, `typescript-eslint` could add validation to warn users when they accidentally enable an ESLint rule that `typescript-eslint` extends:

```js
const extensionRules = ['class-methods-use-this', 'default-param-last', ...];

const extensionRuleValidator = {
    validate*(filePath, fileConfig, eslintOptions) {
        for (const extensionRuleName of extensionRules) {
            if (fileConfig[extensionRuleName]?.[0] !== 0) {
                yield {
                    severity: 'warn',
                    key: ['rules', extensionRuleName, 0],
                    message: `The typescript-eslint rule '@typescript-eslint/${extensionRuleName}' extends and replaces the built-in ESLint rule '${extensionRuleName}'. It is advised that you the typescript-eslint rule instead.`
                }
            }
        }
    }
};
```

This is obviously a very powerful extension point that allows wide-ranging behaviour changes and analysis, so it would need to be designed and used with care, but I think it gives the community a great jumping off place to customize ESLint's behaviour and guide/restrict end-users without having to make modifications to ESLint core.

There is a question of how this might interact with directives in files - that's a complication I think can be worked out at RFC if this idea is considered for moving forward.

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
It's an interesting idea, but not a direction we want to head in at this point. Configuration is the most complex part of ESLint and allowing people to modify a calculated config, potentially in an invalid way, just adds more complexity to it.

There are currently enough ways to mix, match, and modify configs because the config file is JavaScript. You can already pretty easily implement something like a default severity that preprocesses your config before ESLint receives it. The others you mention are also more or less possible right now with a bit of creativity.

I don't think there's a big need to do anything further at this point.
